### DISCUSSION TITLE
Why records with property named Id needs to be partial?

### DISCUSSION THREAD

**Discussion by takla21** [unoplatform/uno#11768] [Q&A]
When we try to build an uno project on any platforms with a record like this

```cs
public record Entity
{
    public string Id { get; init; }
}
```

it fails with this error 
```
error KE0001: The record 'Entity' is eligible to IKeyEquatable generation (due to 'Id') but is not partial. Either make it partial (recommended), either disable implicit IKeyEquality generation using [ImplicitKeys(IsEnabled = false)] on the record itself or on the whole assembly (not recommended).
```

Here are the list of uno packages we reference on our project
```
<PackageReference Include="Uno.Core.Extensions.Logging.Singleton" Version="4.0.1" />
<PackageReference Include="Uno.Microsoft.Xaml.Behaviors.Interactivity.WinUI" Version="2.3.1-uno.2" />
<PackageReference Include="Uno.Microsoft.Xaml.Behaviors.WinUI.Managed" Version="2.3.1-uno.2" />
<PackageReference Include="Uno.Material.WinUI" Version="2.4.1" />
<PackageReference Include="Uno.Toolkit.WinUI" Version="2.4.2" />
<PackageReference Include="Uno.Toolkit.WinUI.Material" Version="2.4.2" />
<PackageReference Include="Uno.SourceGenerationTasks" Version="4.2.0" />
<PackageReference Include="Uno.Injectable" Version="2.0.0-dev.10" />
<PackageReference Include="Uno.CodeGen" Version="2.0.0-dev.10" />
```

**Comment by jeromelaban** [comment]
@carldebilly may know

**Comment by jeanplevesque** [comment]
It's apparently an analyzer that isn't shipped...
https://github.com/unoplatform/uno.extensions/blob/c439932faabf3f3287c5f307e9c01601034b0b6a/src/Uno.Extensions.Core.Generators/AnalyzerReleases.Unshipped.md

**Comment by dr1rrb** [comment]
This is for the `IKeyEquality` generation https://github.com/unoplatform/uno.extensions/blob/main/doc/Overview/KeyEquality/concept.md

As records have their equality generated, we are also generating a "light" version which only compares the ID (that's why it has to be `partial`). The main goal of this is to properly track changes in a collection when you update an immutable record in an immutable list (e.g. `items = items.Replace(items[0], items[0] with {Title = "New title"})`). When comparing only the keys we are able to determine that `items[0]` is just a an updated version of the same entity, so we will raise a "replace" instead of a "remove" + "add" (and `ListView` is able to properly animate that change).

This is natively used by feeds, but if you are not interested by this, you can disable it : https://github.com/unoplatform/uno.extensions/blob/main/doc/Overview/KeyEquality/concept.md#how-to-disable-generation

**Reply by dr1rrb** [reply]
BTW this means that you are also referencing (at least) the [`Uno.Extensions.Core` package](https://www.nuget.org/packages/uno.Extensions.Core) which is not in the list you sent.

**Reply by takla21** [reply]
Ok. However, one thing I don't understand is where this error is coming from? Like, the documentation and the code that @jeanplevesque linked above seems to come from the uno.extensions repo. Which, I would've though that those kind of error pops up only if you have uno.extensions packages referenced explicitly. But, unless `Uno.Core.Extensions.Logging.Singleton` is the reason why we get those error (it would be weird that a logging package is imposing this rule imo),  I don't understand why we're having this.

**Reply by dr1rrb** [reply]
Maybe a transitive reference, but none from the `PackageReference` you have put in the initial message. You must have some other references somewhere (like in the Directory.props file?). You will have a more complete view in VS's SolutionExplorer under the `Dependencies` node:

![image](https://user-images.githubusercontent.com/8635919/226913900-e5040e54-dae3-4abc-a245-a9340eb2656d.png)

**Reply by jeanplevesque** [reply]
Where is the `ImplicitKeyEquality` attribute mentioned in the documentation?

**Reply by jeanplevesque** [reply]
So the attribute [`ImplicitKeyEquality` mentioned in the documentation](https://github.com/unoplatform/uno.extensions/blob/main/doc/Overview/KeyEquality/concept.md#how-to-disable-generation) **doesn't actually exist**.
The real one is called [`ImplicitKeysAttribute`](https://github.com/unoplatform/uno.extensions/blob/c439932faabf3f3287c5f307e9c01601034b0b6a/src/Uno.Extensions.Core/Equality/ImplicitKeysAttribute.cs).

**Reply by dr1rrb** [reply]
https://github.com/unoplatform/uno.extensions/pull/1317

**Reply by jeanplevesque** [reply]
> [unoplatform/uno.extensions#1317](https://github.com/unoplatform/uno.extensions/pull/1317)

That PR doesn't solve the documentation issue because the updated name is still not the right one.
